reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscriber.scala

```diff
@@ -74,85 +83,116 @@ object StreamSubscriber {
             val err = new Error(s"received subscription in invalid state [$o]")
             o -> (F.delay(s.cancel) >> F.raiseError(err))
         }
         case OnNext(a) => {
-          case WaitingOnUpstream(s, r) => Idle(s) -> r.complete(a.some.asRight)
-          case DownstreamCancellation  => DownstreamCancellation -> F.unit
-          case o                       => o -> F.raiseError(new Error(s"received record [$a] in invalid state [$o]"))
+          case WaitingOnUpstream(s) if batchSize > 1   => Receiving(s) -> q.enqueue1(a.some.asRight)
+          case WaitingOnUpstream(s) if batchSize === 1 => Idle(s) -> q.enqueue1(a.some.asRight)
+          case Receiving(s) if counter % batchSize === 0 =>
+            NextBatch(s) -> q.enqueue1(a.some.asRight)
+          case Receiving(s)           => Receiving(s) -> q.enqueue1(a.some.asRight)
+          case DownstreamCancellation => DownstreamCancellation -> F.unit
+          case o                      => o -> F.raiseError(new Error(s"received record [$a] in invalid state [$o]"))
         }
         case OnComplete => {
-          case WaitingOnUpstream(_, r) => UpstreamCompletion -> r.complete(None.asRight)
-          case _                       => UpstreamCompletion -> F.unit
+          case WaitingOnUpstream(_) => UpstreamCompletion -> q.enqueue1(None.asRight)
+          case Receiving(_)         => UpstreamCompletion -> q.enqueue1(None.asRight)
+          case _                    => UpstreamCompletion -> F.unit
         }
         case OnError(e) => {
-          case WaitingOnUpstream(_, r) => UpstreamError(e) -> r.complete(e.asLeft)
-          case _                       => UpstreamError(e) -> F.unit
+          case WaitingOnUpstream(_) =>
+            UpstreamError(e) -> (q.enqueue1(e.asLeft) >> q.enqueue1(None.asRight))
+          case Receiving(_) => UpstreamCompletion -> q.enqueue1(None.asRight)
+          case _            => UpstreamError(e) -> F.unit
         }
         case OnFinalize => {
-          case WaitingOnUpstream(sub, r) =>
-            DownstreamCancellation -> (F.delay(sub.cancel) >> r.complete(None.asRight))
+          case WaitingOnUpstream(sub) =>
+            DownstreamCancellation -> (F.delay(sub.cancel) >> q.enqueue1(None.asRight))
           case Idle(sub) => DownstreamCancellation -> F.delay(sub.cancel)
           case o         => o -> F.unit
         }
-        case OnDequeue(r) => {
-          case Uninitialized          => RequestBeforeSubscription(r) -> F.unit
-          case Idle(sub)              => WaitingOnUpstream(sub, r) -> F.delay(sub.request(1))
-          case err @ UpstreamError(e) => err -> r.complete(e.asLeft)
-          case UpstreamCompletion     => UpstreamCompletion -> r.complete(None.asRight)
-          case o                      => o -> r.complete((new Error(s"received request in invalid state [$o]")).asLeft)
+        case OnDequeue => {
+          case Uninitialized =>
+            RequestBeforeSubscription -> F.unit
```

Comment:
I wish you hadn't changed it from case and expression in the same line to case and expression in different lines. It makes it harder to spot the actual differences.

Reply:
I agree. But that’s what the fs2 auto formatter does.